crates/storage/storage-api/src/chain.rs

```diff
@@ -37,67 +34,47 @@ pub trait BlockBodyWriter<Provider, Body: BlockBody> {
         provider: &Provider,
         block: BlockNumber,
     ) -> ProviderResult<()>;
-}
 
-/// Trait that implements how chain-specific types are written to the storage.
-pub trait ChainStorageWriter<Provider, Primitives: NodePrimitives>:
-    BlockBodyWriter<Provider, <Primitives::Block as Block>::Body>
-{
-}
-impl<T, Provider, Primitives: NodePrimitives> ChainStorageWriter<Provider, Primitives> for T where
-    T: BlockBodyWriter<Provider, <Primitives::Block as Block>::Body>
-{
-}
+    /// Hook to write any custom indices or state that is not covered by other methods.
+    fn write_custom_state(
+        &self,
+        _provider: &Provider,
+        _state: &ExecutionOutcome<N::Receipt>,
+    ) -> ProviderResult<()> {
+        Ok(())
+    }
 
-/// Input for reading a block body. Contains a header of block being read and a list of pre-fetched
-/// transactions.
-pub type ReadBodyInput<'a, B> =
-    (&'a <B as Block>::Header, Vec<<<B as Block>::Body as BlockBody>::Transaction>);
+    /// Hook to revert any [`ChainStorageWriter::write_custom_state`] changes.
+    fn remove_custom_state_above(
```

Comment:
these two traits make a ton of sense and should unblock a lot of functionality

the core limitation then is that implementations must still handle the in memory sections, which I think is fine because we also do this for all of our impls